File: frontend/src/services/puzzle-service.ts
/**
 * Chess Puzzle Service
 * 
 * Service for interacting with puzzle endpoints
 */

import { apiClient } from './api-client';

export interface Puzzle {
  id: number;
  fen: string;
  solution: string;
  difficulty: number;
  description: string;
  theme: string;
  source: string;
  moveCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface PuzzleStats {
  totalPuzzles: number;
  difficultyCount: Record<number, number>;
  themeCount: Record<string, number>;
}

export class PuzzleService {
  private baseUrl: string;

  constructor() {
    this.baseUrl = `${import.meta.env.VITE_API_BASE_URL || 'http://localhost:8080'}/api/puzzles`;
  }

  /**
   * Get all puzzles
   */
  async getAllPuzzles(): Promise<Puzzle[]> {
    const response = await apiClient.get(`${this.baseUrl}`);
    return response;
  }

  /**
   * Get puzzle by ID
   */
  async getPuzzleById(id: number): Promise<Puzzle> {
    const response = await apiClient.get(`${this.baseUrl}/${id}`);
    return response;
  }

  /**
   * Get a random puzzle
   */
  async getRandomPuzzle(): Promise<Puzzle> {
    const response = await apiClient.get(`${this.baseUrl}/random`);
    return response;
  }

  /**
   * Get puzzles by difficulty level
   */
  async getPuzzlesByDifficulty(difficulty: number): Promise<Puzzle[]> {
    const response = await apiClient.get(`${this.baseUrl}/difficulty/${difficulty}`);
    return response;
  }

  /**
   * Get a random puzzle by difficulty
   */
  async getRandomPuzzleByDifficulty(difficulty: number): Promise<Puzzle> {
    const response = await apiClient.get(`${this.baseUrl}/difficulty/${difficulty}/random`);
    return response;
  }

  /**
   * Get puzzles by theme
   */
  async getPuzzlesByTheme(theme: string): Promise<Puzzle[]> {
    const response = await apiClient.get(`${this.baseUrl}/theme/${encodeURIComponent(theme)}`);
    return response;
  }

  /**
   * Get a random puzzle by theme
   */
  async getRandomPuzzleByTheme(theme: string): Promise<Puzzle> {
    const response = await apiClient.get(`${this.baseUrl}/theme/${encodeURIComponent(theme)}/random`);
    return response;
  }

  /**
   * Get puzzles within a difficulty range
   */
  async getPuzzlesByDifficultyRange(min: number, max: number): Promise<Puzzle[]> {
    const response = await apiClient.get(`${this.baseUrl}/difficulty-range?min=${min}&max=${max}`);
    return response;
  }

  /**
   * Get all available themes
   */
  async getAllThemes(): Promise<string[]> {
    const response = await apiClient.get(`${this.baseUrl}/themes`);
    return response;
  }

  /**
   * Get puzzle statistics
   */
  async getPuzzleStats(): Promise<PuzzleStats> {
    const response = await apiClient.get(`${this.baseUrl}/stats`);
    return response;
  }

  /**
   * Search puzzles by multiple criteria
   */
  async searchPuzzles(filters: {
    difficulty?: number;
    theme?: string;
  }): Promise<Puzzle[]> {
    const params = new URLSearchParams();
    
    if (filters.difficulty !== undefined) {
      params.append('difficulty', filters.difficulty.toString());
    }
    
    if (filters.theme) {
      params.append('theme', filters.theme);
    }

    const queryString = params.toString();
    const url = queryString ? `${this.baseUrl}/search?${queryString}` : `${this.baseUrl}/search`;
    
    const response = await apiClient.get(url);
    return response;
  }

  /**
   * Create a new puzzle
   */
  async createPuzzle(puzzleData: {
    fen: string;
    solution: string;
    description: string;
    theme: string;
    difficulty: number;
    moveCount: number;
  }): Promise<Puzzle> {
    const response = await apiClient.post(this.baseUrl, puzzleData);
    return response;
  }

  /**
   * Get puzzles created by the current user
   */
  async getMyPuzzles(): Promise<Puzzle[]> {
    const response = await apiClient.get(`${this.baseUrl}/my-puzzles`);
    return response;
  }

  /**
   * Get puzzles created by a specific user
   */
  async getPuzzlesByUser(userId: number): Promise<Puzzle[]> {
    const response = await apiClient.get(`${this.baseUrl}/by-user/${userId}`);
    return response;
  }

  /**
   * Update an existing puzzle
   */
  async updatePuzzle(id: number, puzzleData: {
    fen: string;
    solution: string;
    description: string;
    theme: string;
    difficulty: number;
    moveCount: number;
  }): Promise<Puzzle> {
    const response = await apiClient.put(`${this.baseUrl}/${id}`, puzzleData);
    return response;
  }

  /**
   * Delete a puzzle
   */
  async deletePuzzle(id: number): Promise<void> {
    await apiClient.delete(`${this.baseUrl}/${id}`);
  }
}

// Singleton instance
let puzzleServiceInstance: PuzzleService | null = null;

export function getPuzzleService(): PuzzleService {
  if (!puzzleServiceInstance) {
    puzzleServiceInstance = new PuzzleService();
  }
  return puzzleServiceInstance;
}

File: backend/src/main/java/com/chesscoach/service/PuzzleService.java
package com.chesscoach.service;

import com.chesscoach.entity.Puzzle;
import com.chesscoach.entity.User;
import com.chesscoach.repository.PuzzleRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class PuzzleService {

    @Autowired
    private PuzzleRepository puzzleRepository;

    /**
     * Get all puzzles
     */
    public List<Puzzle> getAllPuzzles() {
        return puzzleRepository.findAll();
    }

    /**
     * Get puzzle by ID
     */
    public Optional<Puzzle> getPuzzleById(Long id) {
        return puzzleRepository.findById(id);
    }

    /**
     * Get a random puzzle
     */
    public Optional<Puzzle> getRandomPuzzle() {
        return puzzleRepository.findRandomPuzzle();
    }

    /**
     * Get puzzles by difficulty level
     */
    public List<Puzzle> getPuzzlesByDifficulty(Integer difficulty) {
        return puzzleRepository.findByDifficulty(difficulty);
    }

    /**
     * Get a random puzzle by difficulty
     */
    public Optional<Puzzle> getRandomPuzzleByDifficulty(Integer difficulty) {
        return puzzleRepository.findRandomPuzzleByDifficulty(difficulty);
    }

    /**
     * Get puzzles by theme
     */
    public List<Puzzle> getPuzzlesByTheme(String theme) {
        return puzzleRepository.findByTheme(theme);
    }

    /**
     * Get a random puzzle by theme
     */
    public Optional<Puzzle> getRandomPuzzleByTheme(String theme) {
        return puzzleRepository.findRandomPuzzleByTheme(theme);
    }

    /**
     * Get puzzles within a difficulty range
     */
    public List<Puzzle> getPuzzlesByDifficultyRange(Integer minDifficulty, Integer maxDifficulty) {
        return puzzleRepository.findByDifficultyBetween(minDifficulty, maxDifficulty);
    }

    /**
     * Get puzzles by difficulty and theme
     */
    public List<Puzzle> getPuzzlesByDifficultyAndTheme(Integer difficulty, String theme) {
        return puzzleRepository.findByDifficultyAndTheme(difficulty, theme);
    }

    /**
     * Get puzzles created by a specific user
     */
    public List<Puzzle> getPuzzlesByCreator(User user) {
        return puzzleRepository.findByCreatedBy(user);
    }

    /**
     * Get all available themes
     */
    public List<String> getAllThemes() {
        return puzzleRepository.findAllThemes();
    }

    /**
     * Create a new puzzle
     */
    public Puzzle createPuzzle(Puzzle puzzle) {
        return puzzleRepository.save(puzzle);
    }

    /**
     * Update an existing puzzle
     */
    public Optional<Puzzle> updatePuzzle(Long id, Puzzle updatedPuzzle) {
        return puzzleRepository.findById(id)
                .map(puzzle -> {
                    puzzle.setFen(updatedPuzzle.getFen());
                    puzzle.setSolution(updatedPuzzle.getSolution());
                    puzzle.setDifficulty(updatedPuzzle.getDifficulty());
                    puzzle.setDescription(updatedPuzzle.getDescription());
                    puzzle.setTheme(updatedPuzzle.getTheme());
                    puzzle.setSource(updatedPuzzle.getSource());
                    puzzle.setMoveCount(updatedPuzzle.getMoveCount());
                    return puzzleRepository.save(puzzle);
                });
    }

    /**
     * Delete a puzzle
     */
    public boolean deletePuzzle(Long id) {
        if (puzzleRepository.existsById(id)) {
            puzzleRepository.deleteById(id);
            return true;
        }
        return false;
    }

    /**
     * Get statistics about puzzles
     */
    public PuzzleStats getPuzzleStats() {
        PuzzleStats stats = new PuzzleStats();
        stats.setTotalPuzzles(puzzleRepository.count());
        
        // Count puzzles by difficulty
        for (int i = 1; i <= 10; i++) {
            long count = puzzleRepository.countByDifficulty(i);
            stats.addDifficultyCount(i, count);
        }
        
        // Count puzzles by theme
        List<String> themes = getAllThemes();
        for (String theme : themes) {
            long count = puzzleRepository.countByTheme(theme);
            stats.addThemeCount(theme, count);
        }
        
        return stats;
    }

    /**
     * Inner class for puzzle statistics
     */
    public static class PuzzleStats {
        private long totalPuzzles;
        private java.util.Map<Integer, Long> difficultyCount = new java.util.HashMap<>();
        private java.util.Map<String, Long> themeCount = new java.util.HashMap<>();

        public long getTotalPuzzles() {
            return totalPuzzles;
        }

        public void setTotalPuzzles(long totalPuzzles) {
            this.totalPuzzles = totalPuzzles;
        }

        public java.util.Map<Integer, Long> getDifficultyCount() {
            return difficultyCount;
        }

        public void addDifficultyCount(Integer difficulty, Long count) {
            this.difficultyCount.put(difficulty, count);
        }

        public java.util.Map<String, Long> getThemeCount() {
            return themeCount;
        }

        public void addThemeCount(String theme, Long count) {
            this.themeCount.put(theme, count);
        }
    }
}